pkg/providers/tinkerbell/tinkerbell.go

```diff
@@ -448,6 +461,102 @@ func (p *tinkerbellProvider) setMachinesToPXEBoot(ctx context.Context) error {
 	return errorutil.NewAggregate(errs)
 }
 
+// scrubWorkflowsFromTinkerbell removes all workflows in the Tinkerbell stack that feature in hardware by retrieving
+// hardware MAC addresses using tinkerbellHardware. tinkerbellHardware is necessary because MAC addresses aren't
+// available on the Hardware object type.
+func (p *tinkerbellProvider) scrubWorkflowsFromTinkerbell(ctx context.Context, hardware []tinkv1alpha1.Hardware, tinkerbellHardware []*tinkhardware.Hardware) error {
+	workflows, err := p.providerTinkClient.GetWorkflow(ctx)
+	if err != nil {
+		return fmt.Errorf("retrieving workflows: %w", err)
+	}
+
+	hardwareMACLookup, err := createHardwareIDToMACMapping(tinkerbellHardware)
+	if err != nil {
+		return err
+	}
+
+	manifestHardwareMACs, err := createMACSetFromHardwareManifests(hardwareMACLookup, hardware)
+	if err != nil {
+		return err
+	}
+
+	workflowIDs, err := getWorkflowsIDsFromMACs(manifestHardwareMACs, workflows)
+	if err != nil {
+		return err
+	}
+
+	if err := p.providerTinkClient.DeleteWorkflow(ctx, workflowIDs...); err != nil {
+		return fmt.Errorf("could not delete tinkerbell workflow: %v", err)
+	}
+
+	return nil
+}
+
+func createHardwareIDToMACMapping(hardware []*tinkhardware.Hardware) (map[string]string, error) {
+	hardwareMACLookup := make(map[string]string)
+	for _, h := range hardware {
+		if len(h.Network.Interfaces) == 0 {
+			return nil, fmt.Errorf("hardware manifest without interface: hardware ID = '%v'", h.Id)
+		}
+		hardwareMACLookup[h.Id] = h.Network.Interfaces[0].Dhcp.Mac
```

Comment:
what is the point of creating this map? why can't we just create a set of mac addresses here?

Reply:
See https://github.com/aws/eks-anywhere/pull/1717/files/e81272a79ff4255ae6df850f58a3ab83cad3058b#r839844571